graph/coloring/coloring.go

```diff
@@ -0,0 +1,756 @@
+				}
+				break
+			}
+
+		}
+
+		k++
+	}
+
+	return k, colors
+}
+
+// SanSegundo returns an approximate minimal chromatic number of g and a
+// corresponding vertex coloring using the PASS rule with a single run of a
+// greedy coloring algorithm. If a partial coloring is provided the coloring
+// will be consistent with that partial coloring if possible. Otherwise
+// SanSegundo will return ErrInvalidPartialColoring.
+// See San Segundo doi:10.1016/j.cor.2011.10.008 for details of the algorithm.
+func SanSegundo(g graph.Undirected, partial map[int64]int) (k int, colors map[int64]int, err error) {
+	nodes := g.Nodes()
+	n := nodes.Len()
+	if n == 0 {
+		return
+	}
+	partial, ok := isValid(partial, g)
+	if !ok {
+		return -1, nil, ErrInvalidPartialColoring
+	}
+	order := sortBySaturationDegree(nodes, g, partial)
+	order.heuristic = order.pass
+	k, colors = greedyColoringOf(g, order, order.colors)
+	return k, colors, nil
+}
+
+// WelshPowell returns an approximate minimal chromatic number of g and a
+// corresponding vertex coloring using the Welsh and Powell coloring algorithm.
+// If a partial coloring is provided the coloring will be consistent with that
+// partial coloring if possible. Otherwise WelshPowell will return
+// ErrInvalidPartialColoring.
+// See Welsh and Powell doi:10.1093/comjnl/10.1.85 for details of the algorithm.
+func WelshPowell(g graph.Undirected, partial map[int64]int) (k int, colors map[int64]int, err error) {
+	nodes := g.Nodes()
+	n := nodes.Len()
+	if n == 0 {
+		return
+	}
+	partial, ok := isValid(partial, g)
+	if !ok {
+		return -1, nil, ErrInvalidPartialColoring
+	}
+	k, colors = greedyColoringOf(g, sortByDegree(nodes, g), partial)
+	return k, colors, nil
+}
+
+// isValid returns whether the the partial coloring is valid for g. An empty
+// partial coloring is valid. If the partial coloring is not valid, a nil map
+// is returned, otherwise a new non-nil map is returned. If the input partial
+// coloring is nil, a new map is created and returned.
+func isValid(partial map[int64]int, g graph.Undirected) (map[int64]int, bool) {
+	if partial == nil {
+		return make(map[int64]int), true
+	}
+	for id, c := range partial {
+		if g.Node(id) == nil {
+			return nil, false
+		}
+		to := g.From(id)
+		for to.Next() {
+			if oc, ok := partial[to.Node().ID()]; ok && c == oc {
+				return nil, false
+			}
+		}
+	}
+	return clone(partial), true
+}
+
+func clone(colors map[int64]int) map[int64]int {
+	new := make(map[int64]int, len(colors))
+	for id, c := range colors {
+		new[id] = c
+	}
+	return new
+}
+
+// greedyColoring of returns the chromatic number and a graph coloring of g
+// based on the sequential coloring of nodes given by order and stating from
+// the given partial coloring.
+func greedyColoringOf(g graph.Undirected, order graph.Nodes, partial map[int64]int) (k int, colors map[int64]int) {
+	colors = partial
+	constrained := false
+	for _, c := range colors {
+		if c > k {
+			k = c
+			constrained = true
```

Comment:
Yes, this check whether the user has passed in a constraint and whether that constraint will have made the k found through the algorithm an accurate representation of the number of colours. This will happen when len(partial) > 0 and at least one of the colours is not zero. So the two families of cases where this will be false are partial=nil and all coloured nodes in partial are coloured zero. Both of these a rereasonable uses.